### PR TITLE
[CSBindings] Detect situations when type variable bindings could be i…

### PR DESCRIPTION
**Pull request by xedin**
…ncomplete

When `bind param` constraint is associated with a given
type variable a set of bindings collected for it is
potentially incomplete, because binding gathering doesn't
look through related type variables. Which means that
such type variable has to be de-prioritized until
`bind param` constraint is resolved or there is just
nothing else to try, otherwise there is a risk that
solver would skip some of the valid solutions.

This only affects type variable that appears one the
right-hand side of the `bind param` constraint and
represents result type of the closure body, because
left-hand side gets types from overload choices.

Resolves: rdar://problem/45659733

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
Looks like everything is green, @rudkx WDYT?

File: lib/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -2875,9 +2881,14 @@ class ConstraintSystem {
...
-      // If the only difference is default types,
+      // If there is a difference in number of default types,
       // prioritize bindings with fewer of them.
-      return x.NumDefaultableBindings < y.NumDefaultableBindings;
+      if (x.NumDefaultableBindings != y.NumDefaultableBindings)
+        return x.NumDefaultableBindings < y.NumDefaultableBindings;
+
+      // As a last resort, let's check if the bindings are
+      // potentially incomplete, and if so, let's de-prioritize them.
+      return x.PotentiallyIncomplete < y.PotentiallyIncomplete;
```
Comment (rudkx) [inline lib/Sema/ConstraintSystem.h]: Is there a reason that this cannot just be part of the binding score?

Comment (xedin) [reply lib/Sema/ConstraintSystem.h]: Unfortunately yes, it has to be checked after NumDefaultableBindings is not a part of the score at the moment...

I can look into adding both to the score separately

Comment (rudkx) [reply lib/Sema/ConstraintSystem.h]: Ah, I see. Okay then I think this is fine for now and if you want to follow-up with that later that's okay.

**Review by rudkx — APPROVED** [review]

**Pull request merged**